sympy/sets/fancysets.py

```diff
@@ -501,6 +502,15 @@ def doit(self, **kwargs):
             return FiniteSet(*(f(*a) for a in product(*self.base_sets)))
         return self
 
+    def _kind(self):
+        kinds = [arg.kind for arg in self.args if not isinstance(arg, Lambda)]
+        if not kinds:
+            return SetKind()
+        elif all(i == kinds[0] for i in kinds):
+            return kinds[0]
+        else:
+            return SetKind(UndefinedKind)
```

Comment:
This is not correct e.g.:
```python
In [6]: s = ImageSet(Lambda(n, n*eye(2)), {1,2,3})

In [7]: s
Out[7]: 
⎧⎡n  0⎤ │              ⎫
⎨⎢    ⎥ │ n ∊ {1, 2, 3}⎬
⎩⎣0  n⎦ │              ⎭

In [8]: s.doit()
Out[8]: 
⎧⎡1  0⎤  ⎡2  0⎤  ⎡3  0⎤⎫
⎨⎢    ⎥, ⎢    ⎥, ⎢    ⎥⎬
⎩⎣0  1⎦  ⎣0  2⎦  ⎣0  3⎦⎭
```
It should be `return SetKind(self.lamda.expr.kind)`